### ISSUE TITLE
[question] Export package behaviour when choosing to specify a package folder or not

### ISSUE THREAD

**Issue by foundry-markf** [conan-io/conan#7055]
In past recipes we've written, CMake `configure()` has been called from both `build()` and `package()` functions, as described in https://docs.conan.io/en/latest/howtos/cmake_install.html, to do the CMake build in the recipe's `build()` and CMake install in the recipe's `package()` methods.

I've been experimenting to see if a CMake install target could be executed in the Conan `build()` method. This would be a simplification of current recipes, so that CMake `configure()` doesn't need to be executed _twice_.

At first, it looked promising. But I noticed different behaviour in `export-pkg`, depending on _how_ a Conan package is built.

I made a simple repro of what I've seen on this gist: https://gist.github.com/foundry-markf/df06542041a21c14326e379eb86dabc8
I tested this against Conan 1.25.2 on Windows 10. (Assumes CMake 3+ and Ninja)

The goal of the repro is to use a CMake install target to copy a file into the package, but doing it all in the recipe's `build()` method. (Real world scenarios also have a C/C++ build and test step.)

* If I use conan create (`conancreate_works.bat`), then all is well and the package is successfully made.
* If I use local workflow, and specify a package folder (`-pf`) (`localworkflow_specifypkgfolder_works.bat`), then the package is successfully made.
* If I use local workflow, but leave all folders as defaults (`localworkflow_defaultfolders_unexpected.bat`), then the datafile is installed locally (in the default cwd/package folder), but no file is copied to the local cache upon `export-pkg`.

There is no `package()` method in the recipe, since CMake install does the work (since conan build has access to the package folder).

I suspect this behaviour is down to the comment in https://docs.conan.io/en/latest/reference/commands/creator/export-pkg.html indicating that:

> If no --package-folder is given, Conan will execute the method package() once and the self.copy(...) functions will copy matching files from the source_folder and build_folder to the corresponding path in the Conan cache (working directory corresponds to the build_folder).

From debugging through `conans\client\cmd\export_pkg.py`, this certainly seems to be the case. There is no information of where the files had been written in a previous local workflow build step, when the package folder is not specified on the command line.

Since it works 2 out of 3 ways, this came as a surprise. Am I coming up against undefined behaviour, and there's no way to avoid multiple CMake `configure()` steps that works in all methods of package creation?

I don't believe there's a mechanism to detect in the `package()` function whether a package_folder had been specified to it?

It's a minor concern, as when recipes are finished, they ought to only use conan create to build, but during development, I would expect local workflow to be used, and this might trip some people up as an edge case.

Thanks!
Mark

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @foundry-markf 

Thanks for your detailed report, including the gist with code.

I think the issue is lying in the ``conan export-pkg`` having 2 different behaviors:

- If no argument is provided, it assumes it is packaging from a ``build_folder`` and a ``source_folder``, that is, the actual ``package`` has not been created yet
- If the ``--package-folder``, ``-pf`` argument is defined, it understand it as the package has already been locally created, and it does a pure copy of it. 

So in your case, the ``unexpected`` should be to work:

```bash
conan remove -f pkgfolder/dev
conan install .
conan build .
conan package .
conan export-pkg . -pf=.
```

In short: once the package has been created, either by calling ``conan package`` or because ``conan build`` does the ``cmake.install()`` and packages, then it is necessary to do ``conan export-pkg . --package-folder=.``

Please let me know if this clarify a bit.

We are also re-designing the CLI for Conan 2.0, we might want to consider this input for the future ``export-pkg`` command (cc/ @czoido )

**Comment by foundry-markf** [comment]
Hi @memsharded,
Many thanks for your reply.

What you wrote makes complete sense, and I agree with it.

I modified the export-pkg command in your suggestion a little, to make it copy just what was installed by CMake, rather than all files and artifacts from beside the recipe:

```
conan export-pkg . -pf=package
```

and that works as expected now. (So it's just mimicking what Conan chooses for a package folder internally when -pf isn't specified on earlier commands.) This is a gotcha we can just document for package developers, as it's an edge case in my mind.

One of my colleagues also had an idea of providing a `package()` implementation, and in it, detecting when `self.package_folder` is in the local cache, but `self.build_folder` is not. If that case is detected, perform a `self.copy` on everything from the local `package` folder to `self.package_folder`. Then there would be no need to modify any of the CLI commands. It would need to be replicated in all recipes that used the single CMake configure approach, so there's that downfall.

I'm happy for this to be closed now, but if you need to keep it open for any further input on the 2.0 redesign, up to you.

Thanks for the info!

Mark

**Comment by memsharded** [comment]
If the question has been addressed, we can close it now. We will keep track of the ideas for Conan 2.0, no problem.

Thanks for the feedback!
